### PR TITLE
[StreamQueryResult] Batched variant of the StreamQueryResult collector

### PR DESCRIPTION
**Pull request by Tishj**
Similar to the regular variant, this result collector works by incrementally materializing the result, blocking and resuming as needed.

Where the batched variant differs is that this result collector is only used when we're doing a [batched scan](https://github.com/duckdb/duckdb/pull/3700), because of this the order in which the chunks are emitted matters.

We'll have multiple workers working on tasks at the same time, but we need to emit the batches in the right order.
For this we'll make decisions based on the `minimum_batch_index`.

## Implementation Details

We make a queue of chunks that are ready to be emitted
Chunks can be added to this queue if they meet one of the following criteria:
- If they are part of the `minimum_batch_index`, this indicates that there is no lower batch that is still being materialized.
- If their associated batch index is lower than the `minimum_batch_index` and the batch is marked as completed.

For chunks that don't meet these criteria, we use a separate buffer.
This buffer gets populated until it reaches the maximum size, at which point all the other tasks are blocked and will only resume execution when the buffer has room to receive more chunks.

### CODE REVIEW

**Comment by Tishj** [comment]
Should probably wait to be reviewed after #11529 is merged, so we can measure the performance improvement

File: src/execution/operator/helper/physical_buffered_batch_collector.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+	auto min_batch_index = lstate.partition_info.min_batch_index.GetIndex();
+	auto new_index = lstate.partition_info.batch_index.GetIndex();
+
+	auto &buffered_data = dynamic_cast<BatchedBufferedData &>(*gstate.buffered_data);
+	buffered_data.CompleteBatch(batch);
+	lstate.current_batch = new_index;
+	// FIXME: this can move from 'other' chunks to 'current' chunks, increasing the 'current_batch_tuple_count'
+	// We might want to block here if 'current_batch_tuple_count' has already reached the threshold
+	// So we don't completely disregard the BUFFER_SIZE we set
+	buffered_data.UpdateMinBatchIndex(min_batch_index);
```
[the remark this answers is not in the document]
Comment (Mytherin) [reply src/execution/operator/helper/physical_buffered_batch_collector.cpp:82]: e.g.


The code in the sink could look a bit like this (taken from PhysicalBatchCopyToFile)
```cpp

	if (!memory_manager.IsMinimumBatchIndex(batch_index)) {
		memory_manager.UpdateMinBatchIndex(state.partition_info.min_batch_index.GetIndex());

		// we are not processing the current min batch index
		// check if we have exceeded the maximum number of unflushed rows
		if (memory_manager.OutOfMemory(batch_index)) {
			// out-of-memory - stop sinking chunks and instead assist in processing tasks for the minimum batch index
				memory_manager.BlockTask(input.interrupt_state);
                return SinkResultType::BLOCKED;
		}
	}

```

There is a `memory_manager.SetMemorySize` - this is used as bytes, but could also be rows, or could be some estimate of bytes as well - whichever works best. Memory is managed using `IncreaseUnflushedMemory` and `ReduceUnflushedMemory`.

[the remark this answers is not in the document]
Comment (Mytherin) [reply src/execution/operator/helper/physical_buffered_batch_collector.cpp:82]: Maybe the `PhysicalBatchInsert` is actually an easier example

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good - some comments from my side

**Comment by Tishj** [comment]
I tried to add an optimization to not require a lock for in-progress batches, and let that data live only in the local state
To my surprise this caused an ordering issue.

After analyzing the code I believe this is caused by UpdateMinBatchIndex and the logic surrounding the movement of data
```c++
		if (batch > min_batch) {
			// This batch is still in progress, can not be fetched from yet
			break;
		}
```
This makes sure that batches that have a batch index that is greater than the min batch does not get moved.
This got me thinking, what happens if we have completed batches but there is a batch (or multiple) in between that isn't completed.
If it's not registered in the global state we would have no idea that this batch even exists, so we would blindly move the data to the read-queue, messing up the order.

I'm pretty sure that avoiding the global state until a batch is finished exacerbated this problem, but it likely still exists in the current state.
If one such batch never got appended to because it was constantly blocked, then the global state is also not aware of its existence and we run into this same problem.

**Comment by Tishj** [comment]
Actually, that makes no sense?
If the batch would not be completed, the new "minimum batch" could(should) not be greater than that batch

**Comment by Mytherin** [comment]
The minimum batch should always be lower than any batch that has been handed out, or will be handed out in the future. The wrong result there seems unrelated to this as the rows seems to be in the right order, just rows have been dropped?

**Comment by Tishj** [comment]
First I was writing the minimum batch directly to the read queue, removing that fixed the issues I was having.

I think at smaller buffer sizes this is a bit of a hack, because we will still *check* the `current_batch_tuple_count` but it does not increase that counter even for the min_batch.
So we'll cause the `other_batches_tuple_count` to skyrocket until the minimum batch is finished, at which point it (and batches that were also finished in the mean time) is moved to the read queue, increasing `current_batch_tuple_count`, so now the next min batch *is* blocked until the read queue is empty enough for it to continue

**Comment by Tishj** [comment]
Even with this diff we're not getting equivalent behavior
```patch
diff --git a/src/include/duckdb/main/buffered_data/batched_buffered_data.hpp b/src/include/duckdb/main/buffered_data/batched_buffered_data.hpp
index 906ec8c028..e7edb9122c 100644
--- a/src/include/duckdb/main/buffered_data/batched_buffered_data.hpp
+++ b/src/include/duckdb/main/buffered_data/batched_buffered_data.hpp
@@ -68,9 +68,9 @@ private:
 	deque<unique_ptr<DataChunk>> read_queue;
 	idx_t read_queue_capacity;
 	atomic<idx_t> read_queue_byte_count;
+	bool has_new_completes = false;
 
 	map<idx_t, InterruptState> blocked_sinks;
-
 	atomic<idx_t> min_batch;
 	//! Debug variable to verify that order is preserved correctly.
 	idx_t lowest_moved_batch = 0;
diff --git a/src/main/buffered_data/batched_buffered_data.cpp b/src/main/buffered_data/batched_buffered_data.cpp
index 59fcaa1b76..a479c10e34 100644
--- a/src/main/buffered_data/batched_buffered_data.cpp
+++ b/src/main/buffered_data/batched_buffered_data.cpp
@@ -74,6 +74,9 @@ void BatchedBufferedData::MoveCompletedBatches() {
 		if (batch_index > min_batch) {
 			break;
 		}
+		if (!in_progress_batch.completed && batch_index < min_batch.load()) {
+			return;
+		}
 		D_ASSERT(in_progress_batch.completed || batch_index == min_batch.load());
 		// min_batch - took longer than others
 		// min_batch+1 - completed before min_batch
@@ -111,6 +114,7 @@ void BatchedBufferedData::MoveCompletedBatches() {
 		to_remove.pop();
 		buffer.erase(batch_index);
 	}
+	has_new_completes = false;
 }
 
 void BatchedBufferedData::UpdateMinBatchIndex(idx_t min_batch_index) {
@@ -118,7 +122,7 @@ void BatchedBufferedData::UpdateMinBatchIndex(idx_t min_batch_index) {
 
 	auto old_min_batch = min_batch.load();
 	auto new_min_batch = MaxValue(old_min_batch, min_batch_index);
-	if (new_min_batch == min_batch) {
+	if (new_min_batch == min_batch && !has_new_completes) {
 		// No change, early out
 		return;
 	}
@@ -153,6 +157,7 @@ PendingExecutionResult BatchedBufferedData::ReplenishBuffer(StreamQueryResult &r
 
 void BatchedBufferedData::CompleteBatch(idx_t batch) {
 	lock_guard<mutex> lock(glock);
+	has_new_completes = true;
 	auto it = buffer.find(batch);
 	if (it == buffer.end()) {
 		return;
@@ -191,7 +196,6 @@ void BatchedBufferedData::Append(const DataChunk &to_append, idx_t batch) {
 
 	lock_guard<mutex> lock(glock);
 	if (is_minimum) {
-		MoveCompletedBatches();
 		read_queue.push_back(std::move(chunk));
 		read_queue_byte_count += allocation_size;
 	} else {
```

**Comment by Mytherin** [comment]
Thanks! LGTM

**Pull request merged**